src/main/java/org/privacyidea/JSONParser.java

```diff
@@ -0,0 +1,345 @@
+            if (arrMessages != null) {
+                arrMessages.forEach(val -> {
+                    if (val != null) {
+                        response.messages.add(val.getAsString());
+                    }
+                });
+            }
+
+            JsonArray arrChallenges = detail.getAsJsonArray(MULTI_CHALLENGE);
+            if (arrChallenges != null) {
+                for (int i = 0; i < arrChallenges.size(); i++) {
+                    JsonObject challenge = arrChallenges.get(i).getAsJsonObject();
+                    if (TOKEN_TYPE_WEBAUTHN.equals(getString(challenge, TYPE))) {
+                        String webAuthnSignRequest = "";
+                        JsonElement attrElem = challenge.get(ATTRIBUTES);
+                        if (attrElem != null && !attrElem.isJsonNull()) {
+                            JsonElement webauthnElem = attrElem.getAsJsonObject().get(WEBAUTHN_SIGN_REQUEST);
+                            if (webauthnElem != null && !webauthnElem.isJsonNull()) {
+                                webAuthnSignRequest = webauthnElem.toString();
+                            }
+                        }
+                        response.multichallenge.add(new WebAuthn(
+                                getString(challenge, SERIAL),
+                                getString(challenge, MESSAGE),
+                                getString(challenge, TRANSACTION_ID),
+                                webAuthnSignRequest
+                        ));
+                    } else {
+                        response.multichallenge.add(new Challenge(
+                                getString(challenge, SERIAL),
+                                getString(challenge, MESSAGE),
+                                getString(challenge, TRANSACTION_ID),
+                                getString(challenge, TYPE)
+                        ));
+                    }
+                }
+            }
+        }
+        return response;
+    }
+
+    List<TokenInfo> parseTokenInfo(String serverResponse) {
+        List<TokenInfo> ret = new ArrayList<>();
+        JsonObject object;
+        try {
+            object = JsonParser.parseString(serverResponse).getAsJsonObject();
+        } catch (JsonSyntaxException e) {
+            privacyIDEA.error(e);
+            return ret;
+        }
+
+        JsonObject result = object.getAsJsonObject(RESULT);
+        if (result != null) {
+            JsonObject value = result.getAsJsonObject(VALUE);
+            if (value != null) {
+                JsonArray tokens = value.getAsJsonArray(TOKENS);
+                if (tokens != null) {
+                    List<TokenInfo> infos = new ArrayList<>();
+                    tokens.forEach(jsonValue -> infos.add(parseSingleTokenInfo(jsonValue.toString())));
+                    ret = infos;
+                }
+            }
+        }
+        return ret;
+    }
+
+    private TokenInfo parseSingleTokenInfo(String json) {
```

Comment:
In the server naming context, this also would be a tokenobject and not tokeninfo.

These is so much special parsing - how is this necessary or can't this be more generic?